src/aims_ui/api_interaction.py

```diff
@@ -3,43 +3,137 @@
 import requests
 from . import app
 from flask import render_template
+from io import StringIO, BytesIO
 from .models.get_endpoints import get_endpoints
+from .models.get_addresses import get_addresses
 import urllib
+import csv
+
 
 def get_params(all_user_input):
-  """Return a list of aparameters formatted for API header, from class list of inputs"""
-    
+  """Return a list of parameters formatted for API header, from class list of inputs"""
   params = ['verbose=True']
   for param, value in all_user_input.items():
     if not str(value):
       continue
     if (os.getenv('FLASK_ENV') == 'development') and (param == 'epoch'):
       # do not add epoch for testing
       continue
+
+    if type(value) == str:
+      value = value.replace('%', '')
     quoted_param = urllib.parse.quote_plus(str(param))
     quoted_value = urllib.parse.quote_plus(str(value))
-    params.append(quoted_param + '=' + quoted_value)  
-  
+    params.append(quoted_param + '=' + quoted_value)
+
   return '&'.join(params)
 
 
-def api(
-    url,
-    called_from,
-    all_user_input):
+def api(url, called_from, all_user_input):
   """API helper, all pages go through here to interact with API"""
 
-  header = {"Content-Type": "application/json",}
+  header = {
+      "Content-Type": "application/json",
+  }
 
   if (called_from == 'uprn') or (called_from == 'postcode'):
-    url = app.config.get('API_URL') + url + all_user_input.get(called_from)
+    url = app.config.get('API_URL') + url + all_user_input.get(called_from, '')
     params = get_params(all_user_input)
-    r = requests.get(url, params=params, headers=header, )
+    r = requests.get(
```

Comment:
Since these request sections are the same, you can take them out of the if-else. Likewise with the params=get_params.

Because of how Python works, if statements don't create their own name-scope, you can define variables in two separate branch halves and they'll live beyond the end of the branch.

Reply:
Have adjusted this to make it cleaner